File: python/src/ux_channel/arch/project.py
"""Pure project(graph, hello, effects) -> ops[]. Classic floor is permanent."""

from __future__ import annotations

from typing import Any, List, Mapping, Optional

from ux_channel.arch.effects import EffectGraph, Node


def project(
    graph: EffectGraph,
    peer_hello: Optional[Mapping[str, Any]] = None,
    *,
    effects: str = "auto",
) -> List[dict]:
    if effects not in ("auto", "classic"):
        raise ValueError('effects must be "auto" or "classic"')
    hello = dict(peer_hello or {})
    profiles = set(hello.get("profiles") or [])
    features = set(hello.get("features") or [])
    allow_rich = effects == "auto" and (
        "seq" in features
        or "invoke" in features
        or "web.v1" in profiles
        or "agent.v1" in profiles
    )
    classic_only = effects == "classic" or not allow_rich
    # agent.v1 without web.v1: no chrome-only morph/navigate (profiles/agent.v1.md)
    drop_chrome = "agent.v1" in profiles and "web.v1" not in profiles
    out: List[dict] = []
    for node in graph:
        out.extend(_lower(node, classic_only=classic_only, drop_chrome=drop_chrome))
    return out


def _lower(node: Node, *, classic_only: bool, drop_chrome: bool = False) -> List[dict]:
    k = node.kind
    if k == "seq":
        if classic_only:
            ops: List[dict] = []
            for ch in node.children:
                ops.extend(_lower(ch, classic_only=True, drop_chrome=drop_chrome))
            return ops
        return [
            {
                "op": "seq",
                "ops": [
                    o
                    for ch in node.children
                    for o in _lower(ch, classic_only=False, drop_chrome=drop_chrome)
                ],
            }
        ]
    if k == "after":
        ms = int(node.data.get("ms") or 0)
        tid = str(node.data.get("id") or "t")
        body = [
            o
            for ch in node.children
            for o in _lower(ch, classic_only=classic_only, drop_chrome=drop_chrome)
        ]
        if classic_only:
            return body if ms <= 0 else []
        return [{"op": "timer.set", "id": tid, "ms": ms, "ops": body}]
    if k == "invoke":
        if classic_only:
            return [
                o
                for ch in node.children
                for o in _lower(ch, classic_only=True, drop_chrome=drop_chrome)
            ]
        op: dict[str, Any] = {
            "op": "invoke",
            "ref": node.data["ref"],
            "method": node.data["method"],
            "args": node.data.get("args") or {},
        }
        if node.children:
            op["ops"] = [
                o
                for ch in node.children
                for o in _lower(ch, classic_only=False, drop_chrome=drop_chrome)
            ]
        return [op]
    if drop_chrome and k in ("morph", "navigate"):
        return []
    if k == "morph":
        return [{"op": "morph", "target": node.data["target"], "html": node.data["html"]}]
    if k == "toast":
        op = {
            "op": "toast",
            "message": node.data["message"],
            "level": node.data.get("level", "info"),
        }
        if "duration_ms" in node.data:
            op["duration_ms"] = node.data["duration_ms"]
        return [op]
    if k == "navigate":
        return [
            {
                "op": "navigate",
                "href": node.data["href"],
                "replace": bool(node.data.get("replace")),
            }
        ]
    if k == "dispatch":
        op = {"op": "dispatch", "name": node.data["name"]}
        if node.data.get("target"):
            op["target"] = node.data["target"]
        if "detail" in node.data:
            op["detail"] = node.data["detail"]
        return [op]
    return []

File: rust/src/project.rs
//! Pure `project(graph, hello, effects) -> ops[]`. Classic floor is permanent.
//! SPEC: `SPEC/architecture/project.md`

use crate::effects::{EffectGraph, Node};
use serde_json::{json, Value};

pub fn project(graph: &EffectGraph, peer_hello: &Value, effects: &str) -> Result<Vec<Value>, String> {
    if effects != "auto" && effects != "classic" {
        return Err("effects must be \"auto\" or \"classic\"".into());
    }
    let profiles = string_set(peer_hello.get("profiles"));
    let features = string_set(peer_hello.get("features"));
    let allow_rich = effects == "auto"
        && (features.iter().any(|f| f == "seq" || f == "invoke")
            || profiles.iter().any(|p| p == "web.v1" || p == "agent.v1"));
    let classic_only = effects == "classic" || !allow_rich;
    let drop_chrome = profiles.iter().any(|p| p == "agent.v1")
        && !profiles.iter().any(|p| p == "web.v1");
    let mut out = Vec::new();
    for node in graph {
        out.extend(lower(node, classic_only, drop_chrome));
    }
    Ok(out)
}

fn string_set(v: Option<&Value>) -> Vec<String> {
    match v {
        Some(Value::Array(a)) => a
            .iter()
            .filter_map(|x| x.as_str().map(|s| s.to_string()))
            .collect(),
        _ => Vec::new(),
    }
}

fn lower(node: &Node, classic_only: bool, drop_chrome: bool) -> Vec<Value> {
    match node.kind.as_str() {
        "seq" => {
            if classic_only {
                node.children
                    .iter()
                    .flat_map(|ch| lower(ch, true, drop_chrome))
                    .collect()
            } else {
                let kids: Vec<Value> = node
                    .children
                    .iter()
                    .flat_map(|ch| lower(ch, false, drop_chrome))
                    .collect();
                vec![json!({"op": "seq", "ops": kids})]
            }
        }
        "after" => {
            let ms = node
                .data
                .get("ms")
                .and_then(|v| v.as_i64())
                .unwrap_or(0);
            let tid = node
                .data
                .get("id")
                .and_then(|v| v.as_str())
                .unwrap_or("t");
            let body: Vec<Value> = node
                .children
                .iter()
                .flat_map(|ch| lower(ch, classic_only, drop_chrome))
                .collect();
            if classic_only {
                if ms <= 0 {
                    body
                } else {
                    vec![]
                }
            } else {
                vec![json!({"op": "timer.set", "id": tid, "ms": ms, "ops": body})]
            }
        }
        "invoke" => {
            if classic_only {
                return node
                    .children
                    .iter()
                    .flat_map(|ch| lower(ch, true, drop_chrome))
                    .collect();
            }
            let mut op = json!({
                "op": "invoke",
                "ref": node.data.get("ref").cloned().unwrap_or(Value::Null),
                "method": node.data.get("method").cloned().unwrap_or(Value::Null),
                "args": node.data.get("args").cloned().unwrap_or_else(|| json!({})),
            });
            if !node.children.is_empty() {
                let kids: Vec<Value> = node
                    .children
                    .iter()
                    .flat_map(|ch| lower(ch, false, drop_chrome))
                    .collect();
                op["ops"] = json!(kids);
            }
            vec![op]
        }
        "morph" | "navigate" if drop_chrome => vec![],
        "morph" => vec![json!({
            "op": "morph",
            "target": node.data.get("target").cloned().unwrap_or(Value::Null),
            "html": node.data.get("html").cloned().unwrap_or(Value::Null),
        })],
        "toast" => {
            let mut op = json!({
                "op": "toast",
                "message": node.data.get("message").cloned().unwrap_or(Value::Null),
                "level": node.data.get("level").cloned().unwrap_or_else(|| json!("info")),
            });
            if let Some(d) = node.data.get("duration_ms") {
                op["duration_ms"] = d.clone();
            }
            vec![op]
        }
        "navigate" => vec![json!({
            "op": "navigate",
            "href": node.data.get("href").cloned().unwrap_or(Value::Null),
            "replace": node.data.get("replace").and_then(|v| v.as_bool()).unwrap_or(false),
        })],
        "dispatch" => {
            let mut op = json!({
                "op": "dispatch",
                "name": node.data.get("name").cloned().unwrap_or(Value::Null),
            });
            if let Some(t) = node.data.get("target") {
                op["target"] = t.clone();
            }
            if let Some(d) = node.data.get("detail") {
                op["detail"] = d.clone();
            }
            vec![op]
        }
        _ => vec![],
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::effects::{graph, seq, toast};

    #[test]
    fn classic_floor_without_hello() {
        let g = graph([seq([toast("a"), toast("b")])]);
        let ops = project(&g, &json!({}), "auto").unwrap();
        assert_eq!(ops[0]["op"], "toast");
        assert_eq!(ops[1]["op"], "toast");
        assert_eq!(ops.len(), 2);
    }

    #[test]
    fn auto_web_keeps_seq() {
        let g = graph([seq([toast("a"), toast("b")])]);
        let hello = json!({"profiles": ["web.v1"], "features": ["seq"]});
        let ops = project(&g, &hello, "auto").unwrap();
        assert_eq!(ops[0]["op"], "seq");
        assert_eq!(ops[0]["ops"].as_array().unwrap().len(), 2);
    }

    #[test]
    fn classic_mode_flattens_even_with_hello() {
        let g = graph([seq([toast("a"), toast("b")])]);
        let hello = json!({"profiles": ["web.v1"], "features": ["seq"]});
        let ops = project(&g, &hello, "classic").unwrap();
        assert_eq!(ops.len(), 2);
        assert_eq!(ops[0]["op"], "toast");
    }

    #[test]
    fn agent_only_drops_morph() {
        use crate::effects::morph;
        let g = graph([seq([toast("ok"), morph("#x", "<b>no</b>")])]);
        let hello = json!({"profiles": ["agent.v1"], "features": ["seq"]});
        let ops = project(&g, &hello, "auto").unwrap();
        assert_eq!(ops[0]["op"], "seq");
        let kids = ops[0]["ops"].as_array().unwrap();
        assert_eq!(kids.len(), 1);
        assert_eq!(kids[0]["op"], "toast");
    }
}

#[cfg(test)]
mod prop_tests {
    use super::*;
    use crate::effects::{graph, seq, toast};
    use proptest::prelude::*;

    fn kinds(ops: &[Value]) -> Vec<String> {
        let mut out = Vec::new();
        fn walk(list: &[Value], out: &mut Vec<String>) {
            for op in list {
                if let Some(k) = op.get("op").and_then(|v| v.as_str()) {
                    out.push(k.to_string());
                }
                if let Some(Value::Array(kids)) = op.get("ops") {
                    walk(kids, out);
                }
            }
        }
        walk(ops, &mut out);
        out
    }

    proptest! {
        #[test]
        fn classic_never_emits_seq(msg in ".{1,16}") {
            let g = graph([seq([toast(&msg)])]);
            let hello = json!({"profiles": ["web.v1"], "features": ["seq"]});
            let ops = project(&g, &hello, "classic").unwrap();
            prop_assert!(!kinds(&ops).iter().any(|k| k == "seq"));
            prop_assert!(kinds(&ops).iter().any(|k| k == "toast"));
        }

        #[test]
        fn empty_hello_is_floor(msg in ".{1,16}") {
            let g = graph([seq([toast(&msg)])]);
            let ops = project(&g, &json!({}), "auto").unwrap();
            prop_assert!(!kinds(&ops).iter().any(|k| k == "seq"));
        }
    }
}
